sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/MergeScalarSubqueries.scala

```diff
@@ -0,0 +1,357 @@
+
+/**
+ * This rule tries to merge multiple non-correlated [[ScalarSubquery]]s to compute multiple scalar
+ * values once.
+ *
+ * The process is the following:
+ * - While traversing through the plan each [[ScalarSubquery]] plan is tried to merge into the cache
+ *   of already seen subquery plans. If merge is possible then cache is updated with the merged
+ *   subquery plan, if not then the new subquery plan is added to the cache.
+ *   During this first traversal each [[ScalarSubquery]] expression is replaced to a temporal
+ *   [[ScalarSubqueryReference]] reference pointing to its cached version.
+ *   The cache uses a flag to keep track of if a cache entry is a result of merging 2 or more
+ *   plans, or it is a plan that was seen only once.
+ *   Merged plans in the cache get a "Header", that contains the list of attributes form the scalar
+ *   return value of a merged subquery.
+ * - A second traversal checks if there are merged subqueries in the cache and builds a `WithCTE`
+ *   node from these queries. The `CTERelationDef` nodes contain the merged subquery in the
+ *   following form:
+ *   `Project(Seq(CreateNamedStruct(name1, attribute1, ...) AS mergedValue), mergedSubqueryPlan)`
+ *   and the definitions are flagged that they host a subquery, that can return maximum one row.
+ *   During the second traversal [[ScalarSubqueryReference]] expressions that pont to a merged
+ *   subquery is either transformed to a `GetStructField(ScalarSubquery(CTERelationRef(...)))`
+ *   expression or restored to the original [[ScalarSubquery]].
+ *
+ * Eg. the following query:
+ *
+ * SELECT
+ *   (SELECT avg(a) FROM t),
+ *   (SELECT sum(b) FROM t)
+ *
+ * is optimized from:
+ *
+ * == Optimized Logical Plan ==
+ * Project [scalar-subquery#242 [] AS scalarsubquery()#253,
+ *          scalar-subquery#243 [] AS scalarsubquery()#254L]
+ * :  :- Aggregate [avg(a#244) AS avg(a)#247]
+ * :  :  +- Project [a#244]
+ * :  :     +- Relation default.t[a#244,b#245] parquet
+ * :  +- Aggregate [sum(a#251) AS sum(a)#250L]
+ * :     +- Project [a#251]
+ * :        +- Relation default.t[a#251,b#252] parquet
+ * +- OneRowRelation
+ *
+ * to:
+ *
+ * WithCTE
+ * :- CTERelationDef 0
+ * :  +- Project [named_struct(avg(a), avg(a)#247, sum(a), sum(a)#250L) AS mergedValue#260]
+ * :     +- Aggregate [avg(a#244) AS avg(a)#247, sum(a#244) AS sum(a)#250L]
+ * :        +- Project [a#244]
+ * :           +- Relation default.t[a#244,b#245] parquet
+ * +- Project [scalar-subquery#242 [].avg(a) AS scalarsubquery()#253,
+ *             scalar-subquery#243 [].sum(a) AS scalarsubquery()#254L]
+ *    :  :- CTERelationRef 0, true, [mergedValue#260], true
+ *    :  +- CTERelationRef 0, true, [mergedValue#260], true
+ *    +- OneRowRelation
+ */
+object MergeScalarSubqueries extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    plan match {
+      case s: Subquery => s.copy(child = extractCommonScalarSubqueries(s.child))
```

Comment:
The entire optimizer will be recursively called to optimize subqueries in the rule `OptimizeSubqueries`. Since this rule transforms subqueries as well, we should not recursively run it when optimizing subqueries. I think we need to skip this rule if the root node is `Subquery`, which indicates that this rule is being called for subqueires.

Reply:
I'm not sure I get your point. This rule tries to merge subqueries (already optimized by `OptimizeSubqueries`) in a given query/subquery plan. The rule doesn't recurse to nested subqueries.
The test `Merge non-correlated scalar subqueries in a subquery` in https://github.com/apache/spark/pull/32298/files#diff-adc3f627771dab40517626755090a97ce8d6ab3cec413b8fb88627e7049414e0R2055-R2062` would fail if we disable optimizing subqueries here.

Also, the 2 subqueries in this query are not merged: 
```
SELECT
  (SELECT avg(key) FROM testData),
  (
    SELECT
      SUM(
        (SELECT sum(key) FROM testData)
      )
    FROM testData
  )
```
We could merge them, but probably it would be better to do it in follow-up PR...

Reply:
Ah I see, we use `transformAllExpressionsWithPruning`, which means we only merge subqueries at the same level.

Reply:
Now we are only allowed to put `WithCTE` in the root node of the main query, I think we can make the change in this PR now, to merge scalar queries at all the levels (which I think is clearer). We just need to do a bottom-up traversal (inner-most subqueries should be put in the cache first).

If we want a simple version as a start, we can only merge subqueries at the first level.

Reply:
I changed the implementation in https://github.com/apache/spark/pull/32298/commits/c64373b4722bfe41406b555f8cb501c47c0fbd67 and fixed subquery merging from different levels/parents in https://github.com/apache/spark/pull/32298/commits/f93283d3faa8dbd365b0a66b5f7b7ca57bb87ce7 + https://github.com/apache/spark/pull/32298/commits/c26858099a8274548ad8017ed96247cef35e9167.

Reply:
I've simplified `removeReferences()` in https://github.com/apache/spark/pull/32298/commits/1c4d14b8d759b57cca065d316337c7c9cc4d7cb7 and fixed the failing tests.